tasks/ruby.rb

```diff
@@ -5,15 +5,22 @@
 
 class Facts < TaskHelper
   def facter_executable
-    install_path = File.join(File.dirname(RbConfig.ruby), 'facter')
+    exe_path = File.join(File.dirname(RbConfig.ruby), 'facter.exe')
+    bat_path = File.join(File.dirname(RbConfig.ruby), 'facter.bat')
+    ruby_path = File.join(File.dirname(RbConfig.ruby), 'facter')
 
-    # Fall back to PATH lookup if puppet-agent isn't installed
-    if File.exist?(install_path)
-      # Paths with spaces must be quoted on Windows, which means slashes need escaping
-      Gem.win_platform? ? "\"#{install_path}\"" : install_path
-    else
-      'facter'
+    if Gem.win_platform?
+      if File.exist?(exe_path)
+        return "\"#{exe_path}\""
+      elsif File.exist?(bat_path)
+        return "\"#{bat_path}\""
+      end
+    elsif File.exist?(ruby_path)
```

Comment:
This still won't use `ruby_path` on Windows systems that have that path, since any win_platform will only check for the exe_path and bat_path. It sounds like the `ruby_path` is valid in Puppet 7. The `win_platform?` bit shouldn't be necessary if we just check all three, since the exe and bat files shouldn't exist on non-Windows platforms.

```
  def facter_executable
    exe_path = File.join(File.dirname(RbConfig.ruby), 'facter.exe')
    bat_path = File.join(File.dirname(RbConfig.ruby), 'facter.bat')
    ruby_path = File.join(File.dirname(RbConfig.ruby), 'facter')

    if File.exist?(exe_path)
      "\"#{exe_path}\""
    elsif File.exist?(bat_path)
      "\"#{bat_path}\""
    elsif File.exist?(ruby_path)
      ruby_path
    else
      # Fall back to PATH lookup if known path does not exist
      'facter'
    end 
  end 
```

Reply:
Sorry, my original example only worked because it checked the platform *and* path at the same time, which means if the path didn't exist it would still use `ruby_path` regardless of the platform. This checks just the platform first, which means the Windows system will never reach the next `elsif`.

Reply:
That is the issue that this  PR is trying to fix.
On Window `ruby_path` is never going to be an executable file. You cannot run files that don't have `.bat`, `.exe`, `.cmd` or `.com`
This works on Facter as a gem + Agent 6 because, on top of `ruby_path` we ship another file containing the above-mentioned extensions in the `RbConfig.ruby` folder.
On Agent 7 this is not the case. The following file exists: `C:\Program Files\Puppet Labs\Puppet\puppet\bin\facter` but there is no other file with the `.bat`, `.exe`, `.cmd` or `.com` extension

The code snipped fails on Agent 7 with the following exception:
 ```
=> :facter_executable
irb(main):019:0> facter_executable
=> "\"C:/Program Files/Puppet Labs/Puppet/puppet/bin/facter\""
irb(main):017:0> require 'open3'
=> true
irb(main):018:0> stdout, stderr, status = Open3.capture3("#{facter_executable} -v")
Traceback (most recent call last):
        8: from C:/Program Files/Puppet Labs/Puppet/puppet/bin/irb.bat:31:in `<main>'
        7: from C:/Program Files/Puppet Labs/Puppet/puppet/bin/irb.bat:31:in `load'
        6: from C:/Program Files/Puppet Labs/Puppet/puppet/lib/ruby/gems/2.7.0/gems/irb-1.2.3/exe/irb:11:in `<top (required)>'
        5: from (irb):18
        4: from C:/Program Files/Puppet Labs/Puppet/puppet/lib/ruby/2.7.0/open3.rb:281:in `capture3'
        3: from C:/Program Files/Puppet Labs/Puppet/puppet/lib/ruby/2.7.0/open3.rb:101:in `popen3'
        2: from C:/Program Files/Puppet Labs/Puppet/puppet/lib/ruby/2.7.0/open3.rb:213:in `popen_run'
        1: from C:/Program Files/Puppet Labs/Puppet/puppet/lib/ruby/2.7.0/open3.rb:213:in `spawn'
Errno::ENOEXEC (Exec format error - "C:/Program Files/Puppet Labs/Puppet/puppet/bin/facter" -v)
```

Reply:
Ohhh, I see, I misunderstood. Thanks!!